src/main/java/org/opensearch/ad/transport/AnomalyResultTransportAction.java

```diff
@@ -437,110 +442,140 @@ private void handleException(Exception e) {
             long dataEndTime = request.getEnd() - delayMillis;
 
-            // HC logic starts here
-            if (anomalyDetector.isMultientityDetector()) {
-                Optional<Exception> previousException = stateManager.fetchExceptionAndClear(adID);
-
-                if (previousException.isPresent()) {
-                    Exception exception = previousException.get();
-                    LOG.error("Previous exception of {}: {}", adID, exception);
-                    if (exception instanceof EndRunException) {
-                        EndRunException endRunException = (EndRunException) exception;
-                        if (endRunException.isEndNow()) {
-                            listener.onFailure(exception);
-                            return;
-                        }
-                    }
-                }
-
-                // assume request are in epoch milliseconds
-                long nextDetectionStartTime = request.getEnd() + (long) (anomalyDetector.getDetectorIntervalInMilliseconds()
-                    * intervalRatioForRequest);
-
-                CompositeRetriever compositeRetriever = new CompositeRetriever(
-                    dataStartTime,
-                    dataEndTime,
+            adTaskManager
+                .initRealtimeTaskCacheAndCleanupStaleCache(
+                    adID,
                     anomalyDetector,
-                    xContentRegistry,
-                    client,
-                    nextDetectionStartTime,
-                    settings,
-                    maxEntitiesPerInterval,
-                    pageSize
+                    transportService,
+                    ActionListener
+                        .runAfter(
+                            initRealtimeTaskCacheListener(adID),
+                            () -> executeAnomalyDetection(listener, adID, request, anomalyDetector, dataStartTime, dataEndTime)
+                        )
                 );
 
-                PageIterator pageIterator = null;
+            // executeAnomalyDetection(listener, adID, request, anomalyDetector, dataStartTime, dataEndTime);
+        }, exception -> handleExecuteException(exception, listener, adID));
+    }
 
-                try {
-                    pageIterator = compositeRetriever.iterator();
-                } catch (Exception e) {
-                    listener
-                        .onFailure(
-                            new EndRunException(anomalyDetector.getDetectorId(), CommonErrorMessages.INVALID_SEARCH_QUERY_MSG, e, false)
-                        );
-                    return;
+    private ActionListener<Boolean> initRealtimeTaskCacheListener(String detectorId) {
+        return ActionListener.wrap(r -> {
+            if (r) {
+                LOG.debug("Realtime task cache initied for detector {}", detectorId);
+            }
+        }, e -> LOG.error("Failed to init realtime task cache for " + detectorId, e));
+    }
+
+    private void executeAnomalyDetection(
```

Comment:
Just wrap old logic in this method. The old realtime job run logic still the same.